File: python/models/seir_model.py
"""
SEIR Model for Malaria Transmission
==================================

This module implements a Susceptible-Exposed-Infected-Recovered (SEIR) model
for malaria transmission, which includes a latent period (exposed state)
that is more realistic for malaria transmission dynamics.

Key features:
- Latent period (exposed state) before becoming infectious
- Age-specific parameters for different malaria strains
- Seasonal variations in transmission
- Treatment and immunity considerations
"""

import numpy as np
import matplotlib.pyplot as plt
from scipy.integrate import odeint
import pandas as pd
from typing import Tuple, Dict, List, Optional
import seaborn as sns


class MalariaSEIRModel:
    """
    SEIR model for malaria transmission with malaria-specific parameters.
    """
    
    def __init__(self, 
                 beta: float = 0.3,      # Transmission rate (per day)
                 sigma: float = 0.1,     # Incubation rate (1/latent period)
                 gamma: float = 0.05,    # Recovery rate (per day)
                 mu: float = 0.0001,     # Birth/death rate (per day)
                 N: int = 10000,         # Total population
                 E0: int = 5,            # Initial exposed
                 I0: int = 10,           # Initial infected
                 R0: int = 0,            # Initial recovered
                 seasonal_amplitude: float = 0.2,  # Seasonal variation amplitude
                 seasonal_phase: float = 0.0):     # Seasonal phase offset
        """
        Initialize the malaria SEIR model.
        
        Parameters:
        -----------
        beta : float
            Base transmission rate
        sigma : float
            Incubation rate (1/latent period in days)
        gamma : float
            Recovery rate (1/infectious period in days)
        mu : float
            Birth and death rate
        N : int
            Total population size
        E0 : int
            Initial number of exposed individuals
        I0 : int
            Initial number of infected individuals
        R0 : int
            Initial number of recovered individuals
        seasonal_amplitude : float
            Amplitude of seasonal variation in transmission
        seasonal_phase : float
            Phase offset for seasonal variation
        """
        self.beta = beta
        self.sigma = sigma
        self.gamma = gamma
        self.mu = mu
        self.N = N
        self.S0 = N - E0 - I0 - R0  # Initial susceptible
        self.E0 = E0
        self.I0 = I0
        self.R0 = R0
        self.seasonal_amplitude = seasonal_amplitude
        self.seasonal_phase = seasonal_phase
        
        # Calculate basic reproduction number
        self.R0_basic = self.beta / (self.gamma + self.mu)
        
        # Calculate effective reproduction number (accounting for latent period)
        self.R0_effective = self.beta * self.sigma / ((self.sigma + self.mu) * (self.gamma + self.mu))
        
    def seasonal_transmission_rate(self, t: float) -> float:
        """
        Calculate seasonal transmission rate.
        
        Parameters:
        -----------
        t : float
            Time in days
            
        Returns:
        --------
        float
            Seasonal transmission rate
        """
        # Annual seasonal variation
        seasonal_factor = 1 + self.seasonal_amplitude * np.sin(2 * np.pi * t / 365 + self.seasonal_phase)
        return self.beta * seasonal_factor
    
    def seir_equations(self, y: np.ndarray, t: np.ndarray) -> np.ndarray:
        """
        Define the SEIR differential equations.
        
        Parameters:
        -----------
        y : np.ndarray
            Current state [S, E, I, R]
        t : np.ndarray
            Time points
            
        Returns:
        --------
        np.ndarray
            Derivatives [dS/dt, dE/dt, dI/dt, dR/dt]
        """
        S, E, I, R = y
        
        # Seasonal transmission rate
        beta_t = self.seasonal_transmission_rate(t)
        
        # SEIR equations with birth and death
        dSdt = self.mu * self.N - beta_t * S * I / self.N - self.mu * S
        dEdt = beta_t * S * I / self.N - (self.sigma + self.mu) * E
        dIdt = self.sigma * E - (self.gamma + self.mu) * I
        dRdt = self.gamma * I - self.mu * R
        
        return [dSdt, dEdt, dIdt, dRdt]
    
    def simulate(self, t_span: Tuple[float, float] = (0, 365), 
                 num_points: int = 1000) -> Tuple[np.ndarray, np.ndarray]:
        """
        Simulate the SEIR model over time.
        
        Parameters:
        -----------
        t_span : tuple
            Time span (start, end) in days
        num_points : int
            Number of time points to simulate
            
        Returns:
        --------
        tuple
            (time_points, solution_array) where solution_array has shape (num_points, 4)
            with columns [S, E, I, R]
        """
        t = np.linspace(t_span[0], t_span[1], num_points)
        y0 = [self.S0, self.E0, self.I0, self.R0]
        
        solution = odeint(self.seir_equations, y0, t)
        
        return t, solution
    
    def get_epidemic_metrics(self, solution: np.ndarray) -> Dict[str, float]:
        """
        Calculate key epidemic metrics from simulation results.
        
        Parameters:
        -----------
        solution : np.ndarray
            Solution array from simulation
            
        Returns:
        --------
        dict
            Dictionary containing epidemic metrics
        """
        S, E, I, R = solution.T
        
        # Peak infection
        peak_infection = np.max(I)
        peak_time = np.argmax(I)
        
        # Peak exposed
        peak_exposed = np.max(E)
        
        # Final epidemic size
        final_recovered = R[-1]
        attack_rate = final_recovered / self.N
        
        # Time to peak
        time_to_peak = peak_time
        
        # Latent period statistics
        latent_period = 1 / self.sigma
        
        return {
            'R0_basic': self.R0_basic,
            'R0_effective': self.R0_effective,
            'peak_infection': peak_infection,
            'peak_infection_rate': peak_infection / self.N,
            'peak_exposed': peak_exposed,
            'peak_exposed_rate': peak_exposed / self.N,
            'final_attack_rate': attack_rate,
            'time_to_peak': time_to_peak,
            'latent_period': latent_period,
            'infectious_period': 1 / self.gamma
        }
    
    def plot_results(self, t: np.ndarray, solution: np.ndarray, 
                    save_path: str = None) -> None:
        """
        Plot the SEIR model results.
        
        Parameters:
        -----------
        t : np.ndarray
            Time points
        solution : np.ndarray
            Solution array [S, E, I, R]
        save_path : str, optional
            Path to save the plot
        """
        S, E, I, R = solution.T
        
        fig, axes = plt.subplots(2, 3, figsize=(18, 12))
        
        # Main SEIR plot
        ax1 = axes[0, 0]
        ax1.plot(t, S, 'b-', label='Susceptible', linewidth=2)
        ax1.plot(t, E, 'orange', label='Exposed', linewidth=2)
        ax1.plot(t, I, 'r-', label='Infected', linewidth=2)
        ax1.plot(t, R, 'g-', label='Recovered', linewidth=2)
        ax1.set_xlabel('Time (days)')
        ax1.set_ylabel('Number of individuals')
        ax1.set_title('Malaria SEIR Model Dynamics')
        ax1.legend()
        ax1.grid(True, alpha=0.3)
        
        # Infection and exposure rates
        ax2 = axes[0, 1]
        infection_rate = I / self.N * 100
        exposure_rate = E / self.N * 100
        ax2.plot(t, infection_rate, 'r-', label='Infection Rate', linewidth=2)
        ax2.plot(t, exposure_rate, 'orange', label='Exposure Rate', linewidth=2)
        ax2.set_xlabel('Time (days)')
        ax2.set_ylabel('Rate (%)')
        ax2.set_title('Infection and Exposure Rates')
        ax2.legend()
        ax2.grid(True, alpha=0.3)
        
        # Seasonal transmission rate
        ax3 = axes[0, 2]
        beta_seasonal = [self.seasonal_transmission_rate(time) for time in t]
        ax3.plot(t, beta_seasonal, 'purple', linewidth=2)
        ax3.set_xlabel('Time (days)')
        ax3.set_ylabel('Transmission Rate (β)')
        ax3.set_title('Seasonal Transmission Rate')
        ax3.grid(True, alpha=0.3)
        
        # Phase plane (S vs I)
        ax4 = axes[1, 0]
        ax4.plot(S, I, 'purple', linewidth=2)
        ax4.set_xlabel('Susceptible')
        ax4.set_ylabel('Infected')
        ax4.set_title('Phase Plane (S vs I)')
        ax4.grid(True, alpha=0.3)
        
        # Cumulative cases
        ax5 = axes[1, 1]
        cumulative_cases = R + I
        ax5.plot(t, cumulative_cases, 'orange', linewidth=2)
        ax5.set_xlabel('Time (days)')
        ax5.set_ylabel('Cumulative Cases')
        ax5.set_title('Cumulative Malaria Cases')
        ax5.grid(True, alpha=0.3)
        
        # Effective reproduction number over time
        ax6 = axes[1, 2]
        R_eff = [self.seasonal_transmission_rate(time) * S[i] / self.N / (self.gamma + self.mu) 
                for i, time in enumerate(t)]
        ax6.plot(t, R_eff, 'red', linewidth=2)
        ax6.axhline(y=1, color='black', linestyle='--', alpha=0.7, label='Rₑff = 1')
        ax6.set_xlabel('Time (days)')
        ax6.set_ylabel('Effective Reproduction Number')
        ax6.set_title('Effective Reproduction Number')
        ax6.legend()
        ax6.grid(True, alpha=0.3)
        
        plt.tight_layout()
        
        if save_path:
            plt.savefig(save_path, dpi=300, bbox_inches='tight')
        
        plt.show()
    
    def compare_with_sir(self, t_span: Tuple[float, float] = (0, 365)) -> None:
        """
        Compare SEIR model with equivalent SIR model.
        
        Parameters:
        -----------
        t_span : tuple
            Time span for comparison
        """
        # Import basic SIR model
        from basic_sir import MalariaSIRModel
        
        # Create equivalent SIR model (no latent period)
        sir_model = MalariaSIRModel(
            beta=self.beta,
            gamma=self.gamma,
            mu=self.mu,
            N=self.N,
            I0=self.I0 + self.E0,  # Combine exposed and infected for SIR
            R0=self.R0
        )
        
        # Simulate both models
        t_seir, solution_seir = self.simulate(t_span)
        t_sir, solution_sir = sir_model.simulate(t_span)
        
        # Plot comparison
        fig, axes = plt.subplots(2, 2, figsize=(15, 10))
        
        # Susceptible comparison
        axes[0, 0].plot(t_seir, solution_seir[:, 0], 'b-', label='SEIR', linewidth=2)
        axes[0, 0].plot(t_sir, solution_sir[:, 0], 'b--', label='SIR', linewidth=2)
        axes[0, 0].set_xlabel('Time (days)')
        axes[0, 0].set_ylabel('Susceptible')
        axes[0, 0].set_title('Susceptible Population')
        axes[0, 0].legend()
        axes[0, 0].grid(True, alpha=0.3)
        
        # Infected comparison
        axes[0, 1].plot(t_seir, solution_seir[:, 2], 'r-', label='SEIR', linewidth=2)
        axes[0, 1].plot(t_sir, solution_sir[:, 1], 'r--', label='SIR', linewidth=2)
        axes[0, 1].set_xlabel('Time (days)')
        axes[0, 1].set_ylabel('Infected')
        axes[0, 1].set_title('Infected Population')
        axes[0, 1].legend()
        axes[0, 1].grid(True, alpha=0.3)
        
        # Recovered comparison
        axes[1, 0].plot(t_seir, solution_seir[:, 3], 'g-', label='SEIR', linewidth=2)
        axes[1, 0].plot(t_sir, solution_sir[:, 2], 'g--', label='SIR', linewidth=2)
        axes[1, 0].set_xlabel('Time (days)')
        axes[1, 0].set_ylabel('Recovered')
        axes[1, 0].set_title('Recovered Population')
        axes[1, 0].legend()
        axes[1, 0].grid(True, alpha=0.3)
        
        # Exposed (SEIR only)
        axes[1, 1].plot(t_seir, solution_seir[:, 1], 'orange', label='SEIR Exposed', linewidth=2)
        axes[1, 1].set_xlabel('Time (days)')
        axes[1, 1].set_ylabel('Exposed')
        axes[1, 1].set_title('Exposed Population (SEIR only)')
        axes[1, 1].legend()
        axes[1, 1].grid(True, alpha=0.3)
        
        plt.tight_layout()
        plt.show()
    
    def parameter_sensitivity(self, param_name: str, param_values: List[float],
                            t_span: Tuple[float, float] = (0, 365)) -> pd.DataFrame:
        """
        Perform parameter sensitivity analysis.
        
        Parameters:
        -----------
        param_name : str
            Name of parameter to vary ('beta', 'sigma', 'gamma', 'mu')
        param_values : list
            List of parameter values to test
        t_span : tuple
            Time span for simulation
            
        Returns:
        --------
        pd.DataFrame
            DataFrame with results for each parameter value
        """
        results = []
        
        for value in param_values:
            # Create temporary model with modified parameter
            temp_model = MalariaSEIRModel(
                beta=self.beta if param_name != 'beta' else value,
                sigma=self.sigma if param_name != 'sigma' else value,
                gamma=self.gamma if param_name != 'gamma' else value,
                mu=self.mu if param_name != 'mu' else value,
                N=self.N, E0=self.E0, I0=self.I0, R0=self.R0,
                seasonal_amplitude=self.seasonal_amplitude,
                seasonal_phase=self.seasonal_phase
            )
            
            t, solution = temp_model.simulate(t_span)
            metrics = temp_model.get_epidemic_metrics(solution)
            
            results.append({
                'parameter': param_name,
                'value': value,
                **metrics
            })
        
        return pd.DataFrame(results)


def main():
    """
    Main function to demonstrate the malaria SEIR model.
    """
    print("Malaria SEIR Model Simulation")
    print("=" * 40)
    
    # Create model with malaria-specific parameters
    model = MalariaSEIRModel(
        beta=0.4,                    # Transmission rate
        sigma=0.1,                   # Incubation rate (10-day latent period)
        gamma=0.05,                  # Recovery rate (20-day infectious period)
        mu=0.0001,                   # Birth/death rate
        N=10000,                     # Population of 10,000
        E0=20,                       # Start with 20 exposed
        I0=30,                       # Start with 30 infected
        R0=0,                        # No initially recovered
        seasonal_amplitude=0.3,      # 30% seasonal variation
        seasonal_phase=0.0           # Peak in summer
    )
    
    print(f"Model Parameters:")
    print(f"  Transmission rate (β): {model.beta}")
    print(f"  Incubation rate (σ): {model.sigma}")
    print(f"  Recovery rate (γ): {model.gamma}")
    print(f"  Birth/death rate (μ): {model.mu}")
    print(f"  Population size (N): {model.N}")
    print(f"  Seasonal amplitude: {model.seasonal_amplitude}")
    print(f"  Basic reproduction number (R₀): {model.R0_basic:.2f}")
    print(f"  Effective reproduction number (R₀): {model.R0_effective:.2f}")
    print()
    
    # Simulate the model
    t, solution = model.simulate(t_span=(0, 1000), num_points=1000)
    
    # Calculate and display metrics
    metrics = model.get_epidemic_metrics(solution)
    print("Epidemic Metrics:")
    for key, value in metrics.items():
        if 'rate' in key or 'R0' in key or 'period' in key:
            print(f"  {key}: {value:.3f}")
        else:
            print(f"  {key}: {value:.0f}")
    print()
    
    # Plot results
    model.plot_results(t, solution, save_path="../results/plots/malaria_seir.png")
    
    # Compare with SIR model
    print("Comparing with SIR model...")
    model.compare_with_sir()
    
    # Parameter sensitivity analysis
    print("Parameter Sensitivity Analysis:")
    print("-" * 30)
    
    # Test different incubation rates
    sigma_values = [0.05, 0.1, 0.15, 0.2, 0.25]
    sigma_sensitivity = model.parameter_sensitivity('sigma', sigma_values)
    print("Incubation Rate (σ) Sensitivity:")
    print(sigma_sensitivity[['value', 'R0_effective', 'peak_infection_rate', 'final_attack_rate']].round(3))
    print()
    
    # Test different seasonal amplitudes
    seasonal_values = [0.0, 0.1, 0.2, 0.3, 0.4]
    seasonal_sensitivity = model.parameter_sensitivity('beta', seasonal_values)
    print("Seasonal Amplitude Sensitivity:")
    print(seasonal_sensitivity[['value', 'R0_effective', 'peak_infection_rate', 'final_attack_rate']].round(3))


if __name__ == "__main__":
    main()

File: r/models/seir_model.R
# SEIR Model for Malaria Transmission
# ==================================
#
# This script implements a Susceptible-Exposed-Infected-Recovered (SEIR) model
# for malaria transmission, which includes a latent period (exposed state)
# that is more realistic for malaria transmission dynamics.
#
# Key features:
# - Latent period (exposed state) before becoming infectious
# - Age-specific parameters for different malaria strains
# - Seasonal variations in transmission
# - Treatment and immunity considerations

# Load required libraries
library(deSolve)
library(ggplot2)
library(dplyr)
library(gridExtra)

# Define the MalariaSEIRModel class (using R's S3 system)
MalariaSEIRModel <- function(beta = 0.3,      # Transmission rate (per day)
                            sigma = 0.1,     # Incubation rate (1/latent period)
                            gamma = 0.1,     # Recovery rate (per day)
                            mu = 0.0001,     # Birth/death rate (per day)
                            N = 10000,       # Total population
                            E0 = 5,          # Initial exposed
                            I0 = 10,         # Initial infected
                            R0 = 0,          # Initial recovered
                            seasonal_amplitude = 0.2,  # Seasonal variation amplitude
                            seasonal_phase = 0.0) {    # Seasonal phase offset
  
  # Calculate initial susceptible population
  S0 <- N - E0 - I0 - R0
  
  # Calculate basic reproduction number
  R0_basic <- beta / (gamma + mu)
  
  # Calculate effective reproduction number (accounting for latent period)
  R0_effective <- beta * sigma / ((sigma + mu) * (gamma + mu))
  
  # Create model object
  model <- list(
    beta = beta,
    sigma = sigma,
    gamma = gamma,
    mu = mu,
    N = N,
    S0 = S0,
    E0 = E0,
    I0 = I0,
    R0 = R0,
    seasonal_amplitude = seasonal_amplitude,
    seasonal_phase = seasonal_phase,
    R0_basic = R0_basic,
    R0_effective = R0_effective
  )
  
  class(model) <- "MalariaSEIRModel"
  return(model)
}

# Seasonal transmission rate function
seasonal_transmission_rate <- function(t, model) {
  # Annual seasonal variation
  seasonal_factor <- 1 + model$seasonal_amplitude * sin(2 * pi * t / 365 + model$seasonal_phase)
  return(model$beta * seasonal_factor)
}

# SEIR differential equations
seir_equations <- function(t, y, params) {
  with(as.list(c(y, params)), {
    # Seasonal transmission rate
    beta_t <- seasonal_transmission_rate(t, params)
    
    # SEIR equations with birth and death
    dS <- mu * N - beta_t * S * I / N - mu * S
    dE <- beta_t * S * I / N - (sigma + mu) * E
    dI <- sigma * E - (gamma + mu) * I
    dR <- gamma * I - mu * R
    
    return(list(c(dS, dE, dI, dR)))
  })
}

# Simulate the SEIR model
simulate.MalariaSEIRModel <- function(model, t_span = c(0, 365), num_points = 1000) {
  # Create time vector
  times <- seq(from = t_span[1], to = t_span[2], length.out = num_points)
  
  # Initial conditions
  y0 <- c(S = model$S0, E = model$E0, I = model$I0, R = model$R0)
  
  # Parameters for the differential equation
  params <- list(
    beta = model$beta,
    sigma = model$sigma,
    gamma = model$gamma,
    mu = model$mu,
    N = model$N,
    seasonal_amplitude = model$seasonal_amplitude,
    seasonal_phase = model$seasonal_phase
  )
  
  # Solve the differential equations
  solution <- ode(y = y0, times = times, func = seir_equations, parms = params)
  
  return(solution)
}

# Calculate epidemic metrics
get_epidemic_metrics <- function(model, solution) {
  # Extract the solution data
  S <- solution[, "S"]
  E <- solution[, "E"]
  I <- solution[, "I"]
  R <- solution[, "R"]
  
  # Calculate metrics
  peak_infection <- max(I)
  peak_time <- which.max(I)
  peak_exposed <- max(E)
  final_recovered <- R[length(R)]
  attack_rate <- final_recovered / model$N
  
  # Time to peak
  time_to_peak <- peak_time
  
  # Latent period statistics
  latent_period <- 1 / model$sigma
  
  metrics <- list(
    R0_basic = model$R0_basic,
    R0_effective = model$R0_effective,
    peak_infection = peak_infection,
    peak_infection_rate = peak_infection / model$N,
    peak_exposed = peak_exposed,
    peak_exposed_rate = peak_exposed / model$N,
    final_attack_rate = attack_rate,
    time_to_peak = time_to_peak,
    latent_period = latent_period,
    infectious_period = 1 / model$gamma
  )
  
  return(metrics)
}

# Plot the SEIR model results
plot_results <- function(model, solution, save_path = NULL) {
  # Convert solution to data frame for easier plotting
  df <- as.data.frame(solution)
  df$time <- df$time
  
  # Create plots
  p1 <- ggplot(df, aes(x = time)) +
    geom_line(aes(y = S, color = "Susceptible"), size = 1) +
    geom_line(aes(y = E, color = "Exposed"), size = 1) +
    geom_line(aes(y = I, color = "Infected"), size = 1) +
    geom_line(aes(y = R, color = "Recovered"), size = 1) +
    labs(title = "Malaria SEIR Model Dynamics",
         x = "Time (days)",
         y = "Number of individuals",
         color = "Compartment") +
    theme_minimal() +
    theme(legend.position = "bottom")
  
  # Infection and exposure rates
  df$infection_rate <- df$I / model$N * 100
  df$exposure_rate <- df$E / model$N * 100
  p2 <- ggplot(df, aes(x = time)) +
    geom_line(aes(y = infection_rate, color = "Infection Rate"), size = 1) +
    geom_line(aes(y = exposure_rate, color = "Exposure Rate"), size = 1) +
    labs(title = "Infection and Exposure Rates",
         x = "Time (days)",
         y = "Rate (%)",
         color = "Rate") +
    theme_minimal() +
    theme(legend.position = "bottom")
  
  # Seasonal transmission rate
  df$beta_seasonal <- sapply(df$time, function(t) seasonal_transmission_rate(t, model))
  p3 <- ggplot(df, aes(x = time, y = beta_seasonal)) +
    geom_line(color = "purple", size = 1) +
    labs(title = "Seasonal Transmission Rate",
         x = "Time (days)",
         y = "Transmission Rate (β)") +
    theme_minimal()
  
  # Phase plane (S vs I)
  p4 <- ggplot(df, aes(x = S, y = I)) +
    geom_path(color = "purple", size = 1) +
    labs(title = "Phase Plane (S vs I)",
         x = "Susceptible",
         y = "Infected") +
    theme_minimal()
  
  # Cumulative cases
  df$cumulative_cases <- df$R + df$I
  p5 <- ggplot(df, aes(x = time, y = cumulative_cases)) +
    geom_line(color = "orange", size = 1) +
    labs(title = "Cumulative Malaria Cases",
         x = "Time (days)",
         y = "Cumulative Cases") +
    theme_minimal()
  
  # Effective reproduction number over time
  df$R_eff <- sapply(1:nrow(df), function(i) {
    beta_t <- seasonal_transmission_rate(df$time[i], model)
    return(beta_t * df$S[i] / model$N / (model$gamma + model$mu))
  })
  p6 <- ggplot(df, aes(x = time, y = R_eff)) +
    geom_line(color = "red", size = 1) +
    geom_hline(yintercept = 1, color = "black", linetype = "dashed", alpha = 0.7) +
    labs(title = "Effective Reproduction Number",
         x = "Time (days)",
         y = "Effective Reproduction Number") +
    theme_minimal()
  
  # Combine plots
  combined_plot <- grid.arrange(p1, p2, p3, p4, p5, p6, ncol = 2)
  
  # Save plot if path provided
  if (!is.null(save_path)) {
    ggsave(save_path, combined_plot, width = 15, height = 12, dpi = 300)
  }
  
  return(combined_plot)
}

# Compare with SIR model
compare_with_sir <- function(model, t_span = c(0, 365)) {
  # Create equivalent SIR model (no latent period)
  source("basic_sir.R")
  
  sir_model <- MalariaSIRModel(
    beta = model$beta,
    gamma = model$gamma,
    mu = model$mu,
    N = model$N,
    I0 = model$I0 + model$E0,  # Combine exposed and infected for SIR
    R0 = model$R0
  )
  
  # Simulate both models
  solution_seir <- simulate(model, t_span)
  solution_sir <- simulate(sir_model, t_span)
  
  # Create comparison data frame
  df_comparison <- data.frame(
    time = c(solution_seir[, "time"], solution_sir[, "time"]),
    S = c(solution_seir[, "S"], solution_sir[, "S"]),
    I = c(solution_seir[, "I"], solution_sir[, "I"]),
    R = c(solution_seir[, "R"], solution_sir[, "R"]),
    model = c(rep("SEIR", nrow(solution_seir)), rep("SIR", nrow(solution_sir)))
  )
  
  # Add exposed for SEIR
  df_comparison$E <- c(solution_seir[, "E"], rep(NA, nrow(solution_sir)))
  
  # Plot comparison
  p1 <- ggplot(df_comparison, aes(x = time, y = S, color = model, linetype = model)) +
    geom_line(size = 1) +
    labs(title = "Susceptible Population Comparison",
         x = "Time (days)",
         y = "Susceptible",
         color = "Model", linetype = "Model") +
    theme_minimal()
  
  p2 <- ggplot(df_comparison, aes(x = time, y = I, color = model, linetype = model)) +
    geom_line(size = 1) +
    labs(title = "Infected Population Comparison",
         x = "Time (days)",
         y = "Infected",
         color = "Model", linetype = "Model") +
    theme_minimal()
  
  p3 <- ggplot(df_comparison, aes(x = time, y = R, color = model, linetype = model)) +
    geom_line(size = 1) +
    labs(title = "Recovered Population Comparison",
         x = "Time (days)",
         y = "Recovered",
         color = "Model", linetype = "Model") +
    theme_minimal()
  
  # Exposed (SEIR only)
  df_exposed <- df_comparison[!is.na(df_comparison$E), ]
  p4 <- ggplot(df_exposed, aes(x = time, y = E)) +
    geom_line(color = "orange", size = 1) +
    labs(title = "Exposed Population (SEIR only)",
         x = "Time (days)",
         y = "Exposed") +
    theme_minimal()
  
  comparison_plot <- grid.arrange(p1, p2, p3, p4, ncol = 2)
  
  return(comparison_plot)
}

# Parameter sensitivity analysis
parameter_sensitivity <- function(model, param_name, param_values, t_span = c(0, 365)) {
  results <- data.frame()
  
  for (value in param_values) {
    # Create temporary model with modified parameter
    temp_model <- MalariaSEIRModel(
      beta = ifelse(param_name == "beta", value, model$beta),
      sigma = ifelse(param_name == "sigma", value, model$sigma),
      gamma = ifelse(param_name == "gamma", value, model$gamma),
      mu = ifelse(param_name == "mu", value, model$mu),
      N = model$N,
      E0 = model$E0,
      I0 = model$I0,
      R0 = model$R0,
      seasonal_amplitude = model$seasonal_amplitude,
      seasonal_phase = model$seasonal_phase
    )
    
    # Simulate
    solution <- simulate(temp_model, t_span)
    metrics <- get_epidemic_metrics(temp_model, solution)
    
    # Add to results
    results <- rbind(results, data.frame(
      parameter = param_name,
      value = value,
      R0_basic = metrics$R0_basic,
      R0_effective = metrics$R0_effective,
      peak_infection = metrics$peak_infection,
      peak_infection_rate = metrics$peak_infection_rate,
      final_attack_rate = metrics$final_attack_rate,
      time_to_peak = metrics$time_to_peak
    ))
  }
  
  return(results)
}

# Main function to demonstrate the malaria SEIR model
main <- function() {
  cat("Malaria SEIR Model Simulation\n")
  cat("=============================\n\n")
  
  # Create model with malaria-specific parameters
  model <- MalariaSEIRModel(
    beta = 0.4,                    # Transmission rate
    sigma = 0.1,                   # Incubation rate (10-day latent period)
    gamma = 0.05,                  # Recovery rate (20-day infectious period)
    mu = 0.0001,                   # Birth/death rate
    N = 10000,                     # Population of 10,000
    E0 = 20,                       # Start with 20 exposed
    I0 = 30,                       # Start with 30 infected
    R0 = 0,                        # No initially recovered
    seasonal_amplitude = 0.3,      # 30% seasonal variation
    seasonal_phase = 0.0           # Peak in summer
  )
  
  cat("Model Parameters:\n")
  cat("  Transmission rate (β):", model$beta, "\n")
  cat("  Incubation rate (σ):", model$sigma, "\n")
  cat("  Recovery rate (γ):", model$gamma, "\n")
  cat("  Birth/death rate (μ):", model$mu, "\n")
  cat("  Population size (N):", model$N, "\n")
  cat("  Seasonal amplitude:", model$seasonal_amplitude, "\n")
  cat("  Basic reproduction number (R₀):", round(model$R0_basic, 2), "\n")
  cat("  Effective reproduction number (R₀):", round(model$R0_effective, 2), "\n\n")
  
  # Simulate the model
  solution <- simulate(model, t_span = c(0, 1000), num_points = 1000)
  
  # Calculate and display metrics
  metrics <- get_epidemic_metrics(model, solution)
  cat("Epidemic Metrics:\n")
  cat("  R0_basic:", round(metrics$R0_basic, 3), "\n")
  cat("  R0_effective:", round(metrics$R0_effective, 3), "\n")
  cat("  Peak infection:", round(metrics$peak_infection, 0), "\n")
  cat("  Peak infection rate:", round(metrics$peak_infection_rate, 3), "\n")
  cat("  Peak exposed:", round(metrics$peak_exposed, 0), "\n")
  cat("  Peak exposed rate:", round(metrics$peak_exposed_rate, 3), "\n")
  cat("  Final attack rate:", round(metrics$final_attack_rate, 3), "\n")
  cat("  Time to peak:", round(metrics$time_to_peak, 0), "days\n")
  cat("  Latent period:", round(metrics$latent_period, 1), "days\n")
  cat("  Infectious period:", round(metrics$infectious_period, 1), "days\n\n")
  
  # Plot results
  plot_results(model, solution, save_path = "../results/plots/malaria_seir.png")
  
  # Compare with SIR model
  cat("Comparing with SIR model...\n")
  compare_with_sir(model)
  
  # Parameter sensitivity analysis
  cat("Parameter Sensitivity Analysis:\n")
  cat("-------------------------------\n")
  
  # Test different incubation rates
  sigma_values <- c(0.05, 0.1, 0.15, 0.2, 0.25)
  sigma_sensitivity <- parameter_sensitivity(model, "sigma", sigma_values)
  cat("Incubation Rate (σ) Sensitivity:\n")
  print(round(sigma_sensitivity[, c("value", "R0_effective", "peak_infection_rate", "final_attack_rate")], 3))
  cat("\n")
  
  # Test different seasonal amplitudes
  seasonal_values <- c(0.0, 0.1, 0.2, 0.3, 0.4)
  seasonal_sensitivity <- parameter_sensitivity(model, "beta", seasonal_values)
  cat("Seasonal Amplitude Sensitivity:\n")
  print(round(seasonal_sensitivity[, c("value", "R0_effective", "peak_infection_rate", "final_attack_rate")], 3))
  
  return(list(model = model, solution = solution, metrics = metrics))
}

# Run the main function if script is executed directly
if (interactive()) {
  results <- main()
}


